stan/math/prim/prob/wiener4_lccdf.hpp

```diff
@@ -0,0 +1,391 @@
+#ifndef STAN_MATH_PRIM_PROB_WIENER4_LCCDF_HPP
+#define STAN_MATH_PRIM_PROB_WIENER4_LCCDF_HPP
+
+#include <stan/math/prim/prob/wiener4_lcdf.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Log of probability of reaching the upper bound in diffusion process
+ *
+ * @tparam T_a type of boundary
+ * @tparam T_w type of relative starting point
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return log probability to reach the upper bound
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob(const T_a& a, const T_v& v_value, const T_w& w_value) {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto v = -v_value;
+  const auto w = 1 - w_value;
+  if (fabs(v) == 0.0) {
+    return ret_t(log1p(-w));
+  }
+
+  const auto exponent = -2.0 * v * a * (1.0 - w);
+  if (exponent < 0) {
+    return ret_t(log1m_exp(exponent) - log_diff_exp(2 * v * a * w, exponent));
+  } else {
+    return ret_t(log1m_exp(-exponent) - log1m_exp(2 * v * a));
+  }
+}
+
+/**
+ * Calculate parts of the partial derivatives for wiener_prob_grad_a and
+ * wiener_prob_grad_v (on log-scale)
+ *
+ * @tparam T_a type of boundary
+ * @tparam T_w type of relative starting point
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return 'ans' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob_derivative_term(const T_a& a, const T_v& v_value,
+                                        const T_w& w_value) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto exponent_m1 = log1p(-1.1 * 1.0e-8);
+  ret_t ans;
+  const auto v = -v_value;
+  const auto w = 1 - w_value;
+
+  if (v < 0) {
+    const auto exponent_with_1mw = 2.0 * v * a * (1.0 - w);
+    const auto exponent_with_w = 2 * a * v * w;
+    const auto exponent = 2 * a * v;
+
+    if (((exponent_with_1mw >= exponent_m1) || (exponent_with_w >= exponent_m1))
+        || (exponent >= exponent_m1)) {
+      return ret_t(-w);
+    }
+    ans = LOG_TWO + exponent_with_1mw - log1m_exp(exponent_with_1mw);
+    const auto diff_term = log1m_exp(exponent_with_w) - log1m_exp(exponent);
+    const auto log_w = log(w);
+    if (log_w > diff_term) {
+      ans += log_diff_exp(log_w, diff_term);
+      ans = exp(ans);
+    } else {
+      ans += log_diff_exp(diff_term, log_w);
+      ans = -exp(ans);
+    }
+  } else if (v > 0) {
+    const auto exponent_with_1mw = -2.0 * v * a * (1.0 - w);
+    const auto exponent = (-2 * a * v);
+    if ((exponent_with_1mw >= exponent_m1) || (exponent >= exponent_m1)) {
+      return ret_t(-w);
+    }
+    ans = LOG_TWO - log1m_exp(exponent_with_1mw);
+    const auto diff_term
+        = log_diff_exp(exponent_with_1mw, exponent) - log1m_exp(exponent);
```

Comment:
For the two diff terms you have 

```
auto diff_term = log1m_exp(exponent_with_w) - log1m_exp(exponent);
```

and 

```
auto diff_term = log_diff_exp(exponent_with_1mw, exponent) - log1m_exp(exponent);
```

where the second one expands to 

```
auto diff_term = exponent_with_1mw + log1m_exp(exponent - exponent_with_1mw) - log1m_exp(exponent);
```

I'm not really following the second one, can you explain why the v > 0 cut point causes such a change?

Reply:
Here, we calculate the derivative of the probability function that determines the probability to hit the upper OR the lower response boundary. Whether we hit the upper or the lower boundary is determined by the sign of the variable `v`. Therefore, `v` is a cut point in this function.
When you do not only look at `diff_term` but also consider `ans`, then there pops up the missing `exponent_with_1mw` term. 
So we have different functions for which we have to compute the derivative and they depend on the sign of `v`.

Ok?